# ICAEW Data Analytics Certificate Programme Case Study (Analyst pathway)
## Data wrangling

Welcome to the Case Study for the ICAEW Data Analytics Certificate Programme (Analyst pathway). 

We have discussed the background and context and examined the data and techniques to be used in the case study already. Now we will use the skills we have learnt from this course to analyse the sales of the company we may be investing in.


We'll use the Journals and Sales datasets provided. The structure of this notebook is as follows:

- First, we will start off by loading and viewing the datasets.
- We will see that the datasets have a mixture of both numerical and non-numerical features.
- We will see that the datasets have various data quality issues, such as they contain a number of missing entries and duplicates.
- We will use techniques covered in the course to address these issues and prepare the data for analysis.
- Finally, we will append and join any relevant datasets together to wrangle our data into a usable format.

## Loading and viewing the datasets

First, loading and viewing the datasets. The sales dataset and the journals dataset provided are an Excel, and two csv files respectively. We examined how to load Excel and csv datasets within the course in Module 1 of Unit 2: Data Wrangling.

In [6]:
# Import packages required
import pandas as pd
import numpy as np
# Load the sales dataset as sales
sales = pd.read_excel("C:\\Users\\0028773\\Downloads\\Sales (1).xlsx")
# Examine the sales dataset to ensure it has been read in accurately and examine the data contained
print(sales.info())  
display(sales.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51308 entries, 0 to 51307
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Row ID          51308 non-null  int64         
 1   Order ID        51308 non-null  object        
 2   Order Date      51308 non-null  datetime64[ns]
 3   Ship Date       51308 non-null  datetime64[ns]
 4   Ship Mode       51308 non-null  object        
 5   Customer ID     51308 non-null  object        
 6   Customer Name   51308 non-null  object        
 7   Segment         51286 non-null  object        
 8   Postal Code     9998 non-null   float64       
 9   City            51308 non-null  object        
 10  State           51308 non-null  object        
 11  Country         51308 non-null  object        
 12  Region          51308 non-null  object        
 13  Market          51248 non-null  object        
 14  Product ID      51308 non-null  object        
 15  Ca

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Postal Code,City,...,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit,Shipping Cost,Order Priority
0,4042,MX-2015-AB1001539-42353,2019-12-15,2019-12-19,Standard Class,AB-1001539,Aaron Bergman,Consumer,NaN,Apopa,...,FUR-CH-5379,Furniture,Chairs,"Novimex Executive Leather Armchair, Black",610.6000,2,0.00,238.1200,57.833,Medium
1,4041,MX-2015-AB1001539-42353,2019-12-15,2019-12-19,Standard Class,AB-1001539,Aaron Bergman,Consumer,NaN,Apopa,...,OFF-SU-2966,Office Supplies,Supplies,"Acme Box Cutter, High Speed",151.2000,6,0.00,75.6000,10.786,Medium
2,24145,IN-2015-AB1001558-42256,2019-09-09,2019-09-09,Same Day,AB-1001558,Aaron Bergman,Consumer,NaN,Hubli,...,OFF-BI-6383,Office Supplies,Binders,"Wilson Jones Binding Machine, Durable",50.4600,1,0.00,22.6800,10.540,Critical
3,24144,IN-2015-AB1001558-42256,2019-09-09,2019-09-09,Same Day,AB-1001558,Aaron Bergman,Consumer,NaN,Hubli,...,OFF-BI-3737,Office Supplies,Binders,"Cardinal Index Tab, Clear",26.8800,4,0.00,12.0000,6.550,Critical
4,26085,ID-2015-AB1001559-42178,2019-06-23,2019-06-27,Second Class,AB-1001559,Aaron Bergman,Consumer,NaN,Palembang,...,FUR-FU-3935,Furniture,Furnishings,"Deflect-O Door Stop, Erganomic",372.9132,12,0.27,101.8332,53.070,Medium


In [7]:
# Load the first journals dataset as journals1
journals1 = pd.read_csv("C:\\Users\\0028773\\Downloads\\Journals Part 1.csv")
# Examine it to ensure it has been read in accurately and examine the data contained
print(journals1.info())  
display(journals1.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1800 entries, 0 to 1799
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Account      1800 non-null   object 
 1   AccountDesc  1800 non-null   object 
 2   TransDesc    1800 non-null   object 
 3   Debit        1800 non-null   float64
 4   Credit       1800 non-null   float64
 5   Period       1800 non-null   object 
 6   JnlNo        1800 non-null   int64  
 7   JnlDesc      316 non-null    object 
 8   Amount       1800 non-null   float64
 9   JnlPrep      1800 non-null   object 
 10  JnlAuth      1800 non-null   object 
 11  JnlDateTime  1783 non-null   object 
dtypes: float64(3), int64(1), object(8)
memory usage: 168.9+ KB
None


,Account,AccountDesc,TransDesc,Debit,Credit,Period,JnlNo,JnlDesc,Amount,JnlPrep,JnlAuth,JnlDateTime
0,00-80-8033,Provision for Sales Schemes,ZZX,0.00,9668.59,2019-1,1,Reversed By Jnl 2019-9 Journal No. 277,-9668.59,HV09,AS13,01/01/2019 13:04
1,00-10-1002,Provisions - Trade Sales,XXX,9668.59,0.00,2019-1,1,Reversed By Jnl 2019-9 Journal No. 277,9668.59,HV09,AS13,01/01/2019 13:04
2,00-80-8033,Provision for Sales Schemes,ZZX,0.00,291191.30,2019-1,1,Reversed By Jnl 2019-9 Journal No. 277,-291191.30,HV09,AS13,01/01/2019 13:04
3,00-10-1001,Trade Sale Recycle Scheme,ZZZ,291191.30,0.00,2019-1,1,Reversed By Jnl 2019-9 Journal No. 277,291191.30,HV09,AS13,01/01/2019 13:04
4,00-20-2004,Provision for Obselete Inventory,923,0.00,12848.50,2019-1,10,Reversed By Jnl 2019-4 Journal No. 366,-12848.50,DF18,TC01,30/01/2019 09:56


In [8]:
# Load the second journals dataset as journals2
journals2=pd.read_csv("C:\\Users\\0028773\\Downloads\\Journals Part 2.csv")
# Examine it to ensure it has been read in accurately and examine the data contained
print(journals2.info())  
display(journals2.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2201 entries, 0 to 2200
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Account      2201 non-null   object 
 1   AccountDesc  2201 non-null   object 
 2   TransDesc    2201 non-null   object 
 3   Debit        2201 non-null   float64
 4   Credit       2201 non-null   float64
 5   Period       2201 non-null   object 
 6   JnlNo        2201 non-null   int64  
 7   JnlDesc      400 non-null    object 
 8   Amount       2201 non-null   float64
 9   JnlPrep      2201 non-null   object 
 10  JnlAuth      2201 non-null   object 
 11  JnlDateTime  2201 non-null   object 
dtypes: float64(3), int64(1), object(8)
memory usage: 206.5+ KB
None


,Account,AccountDesc,TransDesc,Debit,Credit,Period,JnlNo,JnlDesc,Amount,JnlPrep,JnlAuth,JnlDateTime
0,00-80-8043,Tax Control,XXX,19763.35,0.00,2019-7,253,NaN,19763.35,DF18,AM04,01/07/2019 16:44
1,00-80-8044,Tax Input,XZZX,0.00,19763.35,2019-7,253,NaN,-19763.35,DF18,AM04,01/07/2019 16:44
2,00-80-8043,Tax Control,XXX,0.00,22956.90,2019-7,253,NaN,-22956.90,DF18,AM04,01/07/2019 16:44
3,00-80-8045,Tax Output,ZZXXZ,22956.90,0.00,2019-7,253,NaN,22956.90,DF18,AM04,01/07/2019 16:44
4,00-80-8045,Tax Output,ZZXXZ,40240.30,0.00,2019-7,253,NaN,40240.30,DF18,AM04,01/07/2019 16:44


## Examining and setting data types

As we can see from our first glance at the data, the dataset has a mixture of numerical and non-numerical features. By examining the data types, we can see how Python has interpreted these features and we can change the types of any fields, as required.


In [9]:
# Examine the data types of the sales data
print(sales.dtypes)

Row ID                     int64
Order ID                  object
Order Date        datetime64[ns]
Ship Date         datetime64[ns]
Ship Mode                 object
Customer ID               object
Customer Name             object
Segment                   object
Postal Code              float64
City                      object
State                     object
Country                   object
Region                    object
Market                    object
Product ID                object
Category                  object
Sub-Category              object
Product Name              object
Sales                    float64
Quantity                   int64
Discount                 float64
Profit                   float64
Shipping Cost            float64
Order Priority            object
dtype: object


In [10]:
# Examine the data types of the first journals data
print(journals1.dtypes)

Account         object
AccountDesc     object
TransDesc       object
Debit          float64
Credit         float64
Period          object
JnlNo            int64
JnlDesc         object
Amount         float64
JnlPrep         object
JnlAuth         object
JnlDateTime     object
dtype: object


In [11]:
# Examine the data types of the second journals data
print(journals2.dtypes)

Account         object
AccountDesc     object
TransDesc       object
Debit          float64
Credit         float64
Period          object
JnlNo            int64
JnlDesc         object
Amount         float64
JnlPrep         object
JnlAuth         object
JnlDateTime     object
dtype: object


In [12]:
# Set any data types, if required
# 1. FIXING DATES
# Convert Sales dates (Order Date and Ship Date)
sales['Order Date'] = pd.to_datetime(sales['Order Date'], errors='coerce', format='mixed')
sales['Ship Date'] = pd.to_datetime(sales['Ship Date'], errors='coerce', format='mixed')

# Convert Journal dates 
# journals1 uses the specific format from your image (DD/MM/YYYY)
journals1['JnlDateTime'] = pd.to_datetime(journals1['JnlDateTime'], format='%d/%m/%Y %H:%M', errors='coerce')
journals2['JnlDateTime'] = pd.to_datetime(journals2['JnlDateTime'], errors='coerce', format='mixed')

# 2. FIXING IDENTIFIERS (Converting IDs to Strings)
# This prevents leading zero loss and stops accidental math on ID numbers.
sales_id_cols = ['Order ID', 'Customer ID', 'Product ID', 'Postal Code']
sales[sales_id_cols] = sales[sales_id_cols].astype(str)

# Clean up the '.0' that often appears on Postal Codes when converting from float
sales['Postal Code'] = sales['Postal Code'].str.replace('.0', '', regex=False)

jnl_id_cols = ['Account', 'JnlNo']
journals1[jnl_id_cols] = journals1[jnl_id_cols].astype(str)
journals2[jnl_id_cols] = journals2[jnl_id_cols].astype(str)

## Identifying any data quality issues

We've now examined the data fields and types contained in our data. Next, we should examine the quality of our data. Two of the biggest data quality issues that can affect and distort our analysis are duplicates and missing data. 

Now, examine the datasets provided to identify if there are any duplicate rows or entries in the dataset and if any fields have missing values. 

We have covered how to check for duplicates and for missing values within a dataset in Module 3 of Unit 2: Data Wrangling ‘Duplicates and missing datasets’. If you purchased the learning and certificate Analyst Pathway, we strongly recommend you revisit this content if you are struggling to complete these tasks.

In [13]:
# Examine the number of duplicates in the first journal dataset
num_duplicates = journals1.duplicated().sum()

print(f"Number of duplicate rows in journals1: {num_duplicates}")

Number of duplicate rows in journals1: 15


The following is assessed in the assessment.

The second journals dataset should be the Journals Part 2 dataset. We can examine the duplicates by using the duplicated() command. This produces a boolean indicating whether each row is a duplicate or not.

In [14]:
# Examine the number of duplicates in the second journal dataset
num_duplicates = journals2.duplicated().sum()

print(f"Number of duplicate rows in journals2: {num_duplicates}")

Number of duplicate rows in journals2: 25


##### <b>Please take a note of the number of duplicate journal entries.</b>

<i>Select here to type your answer:25

The following is assessed in the assessment.

We can examine the duplicates by using the duplicated() command. This produces a boolean indicating whether each row is a duplicate or not, with True signifying the sales order is a duplicate.

In [15]:
# Examine the number of duplicates in the sales dataset
num_duplicates = sales.duplicated().sum()

print(f"Number of duplicate rows in sales: {num_duplicates}")

Number of duplicate rows in sales: 18


##### <b>Please take a note of the number of duplicate sales orders.</b>

<i>Select here to type your answer:18

The following is assessed in the assessment.

The first journals dataset is the Journals Part 1 data. We can identify missing values by using the is.na() command, which returns a boolean for each data point on whether it is a missing value or not.

In [16]:
# Examine if there are any missing values in the first journal dataset
journals1.isna().sum()

Account           0
AccountDesc       0
TransDesc         0
Debit             0
Credit            0
Period            0
JnlNo             0
JnlDesc        1484
Amount            0
JnlPrep           0
JnlAuth           0
JnlDateTime      17
dtype: int64

##### <b>Please take a note of the fields which have missing values in the first journals dataset.</b>

<i>Select here to type your answer: JnlDesc and JnlDateTime

In [17]:
# Examine if there are any missing values in the second journal dataset
journals2.isna().sum()

Account           0
AccountDesc       0
TransDesc         0
Debit             0
Credit            0
Period            0
JnlNo             0
JnlDesc        1801
Amount            0
JnlPrep           0
JnlAuth           0
JnlDateTime       0
dtype: int64

The following is assessed in the assessment.

We can identify missing values by using the is.na() command, which returns a boolean for each data point on whether it is a missing value or not.

In [18]:
# Examine if there are any missing values in the sales dataset
sales.isna().sum()

Row ID             0
Order ID           0
Order Date         0
Ship Date          0
Ship Mode          0
Customer ID        0
Customer Name      0
Segment           22
Postal Code        0
City               0
State              0
Country            0
Region             0
Market            60
Product ID         0
Category           0
Sub-Category       0
Product Name       0
Sales              0
Quantity           0
Discount           0
Profit             0
Shipping Cost      0
Order Priority     0
dtype: int64

##### <b>Please take a note of the number of missing values in the Segment column of the sales dataset.</b>

<i>Select here to type your answer: Segment and Market

## Handling and resolving data quality issues

We have identified that there are numerous missing values or duplicate entries in all of the datasets. First we will handle the missing values. 

### Missing values

There are several fields in the data that have missing values to a different extent. The best way to handle these fields depends on what is missing. 

First, let's consider the sales data. We have identified there are missing values for the customer's post code, the order's market and the customer segment. Due to the amount of missing values for Post Code and the fact it is unlikely we could infer them, it may be best to drop Post Code from our sales dataset. However, it may be possible to suggest a suitable value for customer segment and order's market.

While we examined missing values in Module 3 of Unit 2: Data Wrangling, to identify these values we will use the practical skills of subsetting and selecting data that we learnt in Module 2 of Unit 3: Analysing the Data. If you purchased the learning and certificate Analyst Pathway, we recommend you revisit this content  if you are struggling to complete these tasks.

In [21]:
# Drop Post Code from the sales data
sales = sales.drop(columns=['Postal Code'])

In [29]:
# Examine the rows which have missing values for Market in the sales data and examine the regions
missing_market = sales[sales['Market'].isna()]

In [30]:
# Identify the missing value for Market in the sales data using the region field
print("Regions associated with missing Market data:")
print(missing_market[['Region', 'Country', 'Market']])

Regions associated with missing Market data:
                Region      Country Market
190    Central America       Panama    NaN
1127   Central America       Mexico    NaN
1272   Central America       Mexico    NaN
4664   Central America       Mexico    NaN
5260   Central America    Guatemala    NaN
5261   Central America    Guatemala    NaN
8140   Central America  El Salvador    NaN
9223   Central America    Nicaragua    NaN
10182  Central America    Nicaragua    NaN
11154  Central America    Guatemala    NaN
11741  Central America       Mexico    NaN
15356  Central America     Honduras    NaN
15357  Central America     Honduras    NaN
15358  Central America     Honduras    NaN
15359  Central America     Honduras    NaN
16859  Central America       Mexico    NaN
16860  Central America       Mexico    NaN
17581  Central America  El Salvador    NaN
18249  Central America    Nicaragua    NaN
18250  Central America    Nicaragua    NaN
19468  Central America       Mexico    NaN
20085  Ce

In [33]:
# Fill in the missing values for Market in the sales data and save the changes to this field
market_mapping = sales.dropna(subset=['Market']).set_index('Region')['Market'].to_dict()
sales['Market'] = sales['Market'].fillna(sales['Region'].map(market_mapping))
print(f"Missing values in Market: {sales['Market'].isna().sum()}")

Missing values in Market: 0


In [36]:
# Examine the rows that have missing values for Customer Segment in the sales data and create a subset or list of the Customer Names
missing_segment = sales[sales['Segment'].isna()]
missing_segment_customers = missing_segment['Customer Name'].unique().tolist()

# 3. Examine the results
print(f"Number of rows with missing segments: {len(missing_segment)}")
print(f"Customers with missing segments: {missing_segment_customers}")

Number of rows with missing segments: 22
Customers with missing segments: ['Erica Smith', 'Susan Pistek', 'William Brown', 'Zuschuss Donatelli']


In [37]:
# Examine the sales orders for all customers in the list and identify the missing value for Segment
customer_lookup = sales[sales['Customer Name'].isin(missing_segment_customers)]
segment_clues = customer_lookup.dropna(subset=['Segment'])[['Customer Name', 'Segment']].drop_duplicates()
print("Found Segment labels for these customers:")
print(segment_clues)

Found Segment labels for these customers:
            Customer Name   Segment
17037         Erica Smith  Consumer
46880        Susan Pistek  Consumer
50906       William Brown  Consumer
51254  Zuschuss Donatelli  Consumer


In [46]:
# Resolve the missing values for Customer Segment in the sales data and save the changes to this field
# 1. Create a dictionary that maps Customer Name to their Segment
# We use only rows that HAVE a segment to build our "lookup table"
segment_map = sales.dropna(subset=['Segment']).set_index('Customer Name')['Segment'].to_dict()

# 2. Fill the missing values in 'Segment' using the Customer Name as the key
sales['Segment'] = sales['Segment'].fillna(sales['Customer Name'].map(segment_map))

# 3. Verify that there are no missing values left in Segment
print(f"Remaining missing values in Segment: {sales['Segment'].isna().sum()}")

Remaining missing values in Segment: 0


We have identified that several journal entries have missing values for the time and date the journal was posted. We may be able to infer these values from the rest of the journals data.

The following is assessed in the assessment.

We can identify the missing journal date and time by examining the other journal entries from the same journal number. This is because these will have been posted together at the same time.

In [48]:
# Identify the rows of the missing values for Journal Date Time
missing_journal_date_time = journals1[journals1['JnlDateTime'].isna()]
print(f"Total rows with missing JnlDateTime: {len(missing_journal_date_time)}")
display(missing_journal_date_time)

Total rows with missing JnlDateTime: 0


,Account,AccountDesc,TransDesc,Debit,Credit,Period,JnlNo,JnlDesc,Amount,JnlPrep,JnlAuth,JnlDateTime


In [44]:
# Identify the missing values for Journal Date Time
missing_dates = journals1[journals1['JnlDateTime'].isna()]

# View the results
print(f"Number of rows missing a date: {len(missing_dates)}")

Number of rows missing a date: 0


##### <b>Please take a note of the missing value in Journal Date Time.</b>

<i>Select here to type your answer:

In [ ]:
# Fill in the missing values for Journal Date Time 
# 1. Find the timestamp used for Journal No. 23
# We look for rows where JnlNo is '23' and JnlDateTime is NOT null
jnl_23_date = journals1[(journals1['JnlNo'] == '23') & (journals1['JnlDateTime'].notna())]['JnlDateTime'].iloc[0]

# 2. Apply this date to all rows where JnlNo is '23' and the date is missing
journals1.loc[(journals1['JnlNo'] == '23') & (journals1['JnlDateTime'].isna()), 'JnlDateTime'] = jnl_23_date

In [45]:
# Check the missing values have been filled in
print(f"Remaining missing dates for Jnl 23: {journals1[journals1['JnlNo'] == '23']['JnlDateTime'].isna().sum()}")

Remaining missing dates for Jnl 23: 0


### Duplicates

We have identified that there are duplicate entries in all of the datasets.

In the first journals dataset, there is a whole journal that has been duplicated and a few duplicated journal entries. This is also the case for the second journals dataset. Lastly, there are several sales orders that are duplicated in the sales data. To avoid these compromising our analysis, we will remove these duplicates.

We have examined how to drop duplicates in Module 3 of Unit 2: Data Wrangling. If you purchased the learning and certificate Analyst Pathway, we recommend you revisit this content if you are struggling to complete these tasks.

In [49]:
# Remove the duplicates in the first journals data and save this dataset as journals1
journals1 = journals1.drop_duplicates()

# Verify the new row count
print(f"Number of rows remaining in journals1: {len(journals1)}")

Number of rows remaining in journals1: 1785


In [50]:
# Remove the duplicates in the second journals data and save this dataset as journals2
journals2 = journals2.drop_duplicates()

# Verify the new row count
print(f"Number of rows remaining in journals2: {len(journals2)}")

Number of rows remaining in journals2: 2176


In [51]:
# Remove the duplicates in the sales data and save this dataset as sales
sales = sales.drop_duplicates()

# Verify the new row count
print(f"Number of rows remaining in sales: {len(sales)}")

Number of rows remaining in sales: 51290


## Wrangling the data

Now we have resolved our data quality issues, we can wrangle our data into a usable format for our analysis. 

Until now, we have been using three datasets, two of which are journal datasets covering different periods. For our analysis, it is better to use a single journal dataset to allow us to examine and analyse the journals data as a whole.

We practised appending the journals data and performing joins in Module 2 of Unit 2: Data Wrangling. If you purchased the learning and certificate Analyst Pathway, we recommend you revisit this content if you are struggling to complete these tasks.

In [52]:
# Combine the two journals datasets into one dataset called journals
journals = pd.concat([journals1, journals2], ignore_index=True)
print(f"Total rows in combined journals: {len(journals)}")
journals.head()

Total rows in combined journals: 3961


,Account,AccountDesc,TransDesc,Debit,Credit,Period,JnlNo,JnlDesc,Amount,JnlPrep,JnlAuth,JnlDateTime
0,00-80-8033,Provision for Sales Schemes,ZZX,0.00,9668.59,2019-1,1,Reversed By Jnl 2019-9 Journal No. 277,-9668.59,HV09,AS13,2019-01-01 13:04:00
1,00-10-1002,Provisions - Trade Sales,XXX,9668.59,0.00,2019-1,1,Reversed By Jnl 2019-9 Journal No. 277,9668.59,HV09,AS13,2019-01-01 13:04:00
2,00-80-8033,Provision for Sales Schemes,ZZX,0.00,291191.30,2019-1,1,Reversed By Jnl 2019-9 Journal No. 277,-291191.30,HV09,AS13,2019-01-01 13:04:00
3,00-10-1001,Trade Sale Recycle Scheme,ZZZ,291191.30,0.00,2019-1,1,Reversed By Jnl 2019-9 Journal No. 277,291191.30,HV09,AS13,2019-01-01 13:04:00
4,00-20-2004,Provision for Obselete Inventory,923,0.00,12848.50,2019-1,10,Reversed By Jnl 2019-4 Journal No. 366,-12848.50,DF18,TC01,2019-01-30 09:56:00



In addition to the journal datasets provided, we have obtained a dataset containing the employee names for the accounts team responsbile for preparing and authorising the journals. It will be beneficial to include this information in the journal dataset.

In [54]:
# Import the Accounts Team dataset as AccountsTeam
AccountsTeam=pd.read_csv("C:\\Users\\0028773\\Downloads\\Accounts Team Staff.csv")
# Examine the data
print(AccountsTeam.info())  
display(AccountsTeam.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13 entries, 0 to 12
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   EmployeeRef    13 non-null     object
 1   Employee Name  13 non-null     object
dtypes: object(2)
memory usage: 340.0+ bytes
None


,EmployeeRef,Employee Name
0,AC04,Lorelei Ory
1,AH12,Latasha Terpstra
2,JC39,Merrill Benzel
3,ID03,Evalyn Reddout
4,HV09,Lakeisha Testerman


In [68]:
# Join the Accounts Team data to the journals dataset

# First, join on the Journal Preparer

print(journals.info())  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3961 entries, 0 to 3960
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Account          3961 non-null   object        
 1   AccountDesc      3961 non-null   object        
 2   TransDesc        3961 non-null   object        
 3   Debit            3961 non-null   float64       
 4   Credit           3961 non-null   float64       
 5   Period           3961 non-null   object        
 6   JnlNo            3961 non-null   object        
 7   JnlDesc          700 non-null    object        
 8   Amount           3961 non-null   float64       
 9   JnlPrep          3961 non-null   object        
 10  JnlAuth          3961 non-null   object        
 11  JnlDateTime      3961 non-null   datetime64[ns]
 12  Employee Name_x  3961 non-null   object        
 13  EmployeeRef_y    3961 non-null   object        
dtypes: datetime64[ns](1), float64(3), object

In [69]:
# Rename the new column as JnlPreparerName and drop redundant columns EmployeeRef and Employee Name 
journals.rename(columns={'Employee Name_x': 'JnlPreparerName'}, inplace=True)
journals.drop(columns=['EmployeeRef_y', 'Employee Name'], inplace=True, errors='ignore')
print(journals.columns)

Index(['Account', 'AccountDesc', 'TransDesc', 'Debit', 'Credit', 'Period',
       'JnlNo', 'JnlDesc', 'Amount', 'JnlPrep', 'JnlAuth', 'JnlDateTime',
       'JnlPreparerName'],
      dtype='object')


In [70]:
# Next, join on the Journal Authoriser field
journals = journals.merge(AccountsTeam[['EmployeeRef', 'Employee Name']], 
                          left_on='JnlAuth', 
                          right_on='EmployeeRef', 
                          how='left')

In [71]:
# Again rename new column as JnlAuthoriserName and drop redundant columns EmployeeRef and Employee Name 
journals.rename(columns={'Employee Name': 'JnlAuthoriserName'}, inplace=True)
journals.drop(columns=['EmployeeRef'], inplace=True)

In [72]:
# Examine your final data
print(journals.columns)

Index(['Account', 'AccountDesc', 'TransDesc', 'Debit', 'Credit', 'Period',
       'JnlNo', 'JnlDesc', 'Amount', 'JnlPrep', 'JnlAuth', 'JnlDateTime',
       'JnlPreparerName', 'JnlAuthoriserName'],
      dtype='object')


The following is assessed in the assessment.

When dropping duplicates, we need to save the changes to our datasets in order to obtain the correct number of rows.
Similarly, we need to save the changes when appending the journal datasets together, and when we are merging datasets to add the employee names and dropping irrelevant columns.

In [73]:
# Examine the size of the final journal dataset
print(journals.info())  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3961 entries, 0 to 3960
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Account            3961 non-null   object        
 1   AccountDesc        3961 non-null   object        
 2   TransDesc          3961 non-null   object        
 3   Debit              3961 non-null   float64       
 4   Credit             3961 non-null   float64       
 5   Period             3961 non-null   object        
 6   JnlNo              3961 non-null   object        
 7   JnlDesc            700 non-null    object        
 8   Amount             3961 non-null   float64       
 9   JnlPrep            3961 non-null   object        
 10  JnlAuth            3961 non-null   object        
 11  JnlDateTime        3961 non-null   datetime64[ns]
 12  JnlPreparerName    3961 non-null   object        
 13  JnlAuthoriserName  3961 non-null   object        
dtypes: datet

##### <b>Please take a note of the size of the final wrangled journals dataset.</b>

<i>Select here to type your answer: 3961 rows and 14 columns

##### Assessment guidance

You are ready to take the assessment.

Remember, you should have fully completed your task and recorded your answers in Jupyter Notebook before moving on to the assessment. You can keep your Jupyter Notebook open in a separate browser window to refer to as you take the assessment.  

You will receive a score following completion of the assessment. If you have scored below the target mark of 60% for the section, you are recommended to refresh your knowledge in the course content (if purchased) and rework your Jupyter Notebook before re-attempting the assessment. You have a maximum of three assessment attempts.

You should aim to achieve a target score of 60% in each section of the case study. To pass the case study and be awarded the ICAEW Certificate, you are required to achieve a pass mark of 60% overall, averaged over all five sections, so do not be disheartened if you score below 60% in any one section, as a higher score in one or more of the other sections will contribute to the overall pass mark of 60%.

IMPORTANT: When submitting to the assessment portal, please do not navigate away from it until you have submitted all of your answers for that task. In between any of your 3 assessment attempts (but not during an attempt) you may navigate back to the course content, if you have purchased it, to refresh your knowledge and revisit your Jupyter Notebook to rework your analysis.